File: packages/py/src/config_discovery/merge.py
"""Merge semantics - SPEC section 4.

Dynaconf has no switch that produces the combination the spec requires - deep-merged maps *with*
wholesale-replaced lists - so the merge is written here rather than configured there.

The structure deliberately mirrors ``packages/ts/src/merge.ts``. When a fixture disagreement has
to be adjudicated, two independent implementations that are structurally comparable are far
easier to reconcile than two that are merely both correct.
"""

from __future__ import annotations

from typing import Any, Callable

from .types import Layer

#: Python conflates "key not present" with ``None``, and SPEC section 4.4 depends on telling
#: them apart: an absent key is left alone, a ``None`` key is deleted. This sentinel is the
#: difference, and it is the single most likely bug in this file.
_ABSENT = object()


def _merge_into(
    lower: dict[str, Any],
    higher: dict[str, Any],
    *,
    array_merge: str,
    on_warning: Callable[[str], None] | None,
    path: str,
) -> dict[str, Any]:
    out = dict(lower)
    for key, value in higher.items():
        at = key if not path else f"{path}.{key}"

        # SPEC section 4.4.
        if value is None:
            out.pop(key, None)
            continue

        existing = out.get(key, _ABSENT)

        # An explicit isinstance(..., dict) check, not Mapping: a LazySettings or a DynaBox that
        # leaked this far must fail visibly rather than merge as if it were data.
        if isinstance(existing, dict) and isinstance(value, dict):
            out[key] = _merge_into(
                existing, value, array_merge=array_merge, on_warning=on_warning, path=at
            )
            continue

        if isinstance(existing, list) and isinstance(value, list):
            # SPEC section 4.3: replace by default; concat appends and never deduplicates.
            out[key] = [*existing, *value] if array_merge == "concat" else value
            continue

        if existing is not _ABSENT and isinstance(existing, dict) != isinstance(value, dict):
            if on_warning is not None:
                was = "a map" if isinstance(existing, dict) else "a scalar"
                now = "a map" if isinstance(value, dict) else "a scalar"
                on_warning(f"{at}: replacing {was} with {now} (SPEC section 4.2)")

        out[key] = value
    return out


def merge_layers(
    layers: list[Layer],
    *,
    array_merge: str = "replace",
    on_warning: Callable[[str], None] | None = None,
) -> dict[str, Any]:
    """Fold the layers lowest precedence first."""
    result: dict[str, Any] = {}
    for layer in layers:
        if not isinstance(layer.value, dict):
            continue
        result = _merge_into(
            result, layer.value, array_merge=array_merge, on_warning=on_warning, path=""
        )
    return result


def apply_strategy(layers: list[Layer], strategy: str) -> list[Layer]:
    """SPEC section 3.2.

    Under ``first-match`` only the highest-precedence root that contributed a file survives - the
    lower roots are dropped from the merge **and** from ``sources``, because the option means
    "the others were never consulted", not "the others lost".

    The rootless layers - defaults, env, overrides - always survive: the option scopes the file
    layers only.
    """
    if strategy != "first-match":
        return layers
    rooted = [layer.root for layer in layers if layer.root is not None]
    winning_root = rooted[-1] if rooted else None
    return [layer for layer in layers if layer.root is None or layer.root == winning_root]

File: packages/golang/merge.go
package configdiscovery

// Merge semantics - SPEC section 4.
//
// viper.MergeInConfig does not implement it: it merges into a lowercased key store and has its
// own opinion about slices. So the merge is written here, structurally mirroring the TypeScript
// and Python implementations - three comparable implementations make a fixture disagreement far
// cheaper to adjudicate than three merely correct ones.

import "sort"

// Layer is one layer on its way into the merge, with the source it will be reported as.
type Layer struct {
	Value  map[string]any
	Source Source
	// Root is the config directory this layer came from, for first-match. Empty for the layers
	// that belong to no root: defaults, env, overrides.
	Root string
}

type mergeOptions struct {
	arrayMerge string
	warn       func(string)
}

// deepCopy prevents a layer's map from being aliased into the result and then mutated by a later
// layer - or by the caller.
func deepCopy(value any) any {
	switch typed := value.(type) {
	case map[string]any:
		out := make(map[string]any, len(typed))
		for key, item := range typed {
			out[key] = deepCopy(item)
		}
		return out
	case []any:
		out := make([]any, len(typed))
		for i, item := range typed {
			out[i] = deepCopy(item)
		}
		return out
	default:
		return value
	}
}

func mergeInto(lower, higher map[string]any, o mergeOptions, path string) map[string]any {
	out := make(map[string]any, len(lower)+len(higher))
	for key, value := range lower {
		out[key] = value
	}

	// Ranging yields only the keys the higher layer actually has, which is what keeps absent and
	// null apart: a key that is present with a nil value asks for a delete, a key that is not
	// there at all says nothing. Never test presence with out[key] != nil.
	keys := make([]string, 0, len(higher))
	for key := range higher {
		keys = append(keys, key)
	}
	sort.Strings(keys) // Deterministic warning order; the result is a map either way.

	for _, key := range keys {
		value := higher[key]
		at := key
		if path != "" {
			at = path + "." + key
		}

		// SPEC section 4.4: an explicit null in a higher layer removes the key.
		if value == nil {
			delete(out, key)
			continue
		}

		existing, present := out[key]
		existingMap, existingIsMap := existing.(map[string]any)
		valueMap, valueIsMap := value.(map[string]any)
		if present && existingIsMap && valueIsMap {
			out[key] = mergeInto(existingMap, valueMap, o, at)
			continue
		}

		existingSlice, existingIsSlice := existing.([]any)
		valueSlice, valueIsSlice := value.([]any)
		if present && existingIsSlice && valueIsSlice {
			// SPEC section 4.3: replace by default; concat appends and never deduplicates.
			if o.arrayMerge == "concat" {
				combined := make([]any, 0, len(existingSlice)+len(valueSlice))
				combined = append(combined, existingSlice...)
				combined = append(combined, valueSlice...)
				out[key] = deepCopy(combined)
			} else {
				out[key] = deepCopy(valueSlice)
			}
			continue
		}

		if present && o.warn != nil && existingIsMap != valueIsMap {
			was, now := "a scalar", "a scalar"
			if existingIsMap {
				was = "a map"
			}
			if valueIsMap {
				now = "a map"
			}
			o.warn(at + ": replacing " + was + " with " + now + " (SPEC section 4.2)")
		}
		out[key] = deepCopy(value)
	}
	return out
}

// mergeLayers folds the layers lowest precedence first.
func mergeLayers(layers []Layer, o mergeOptions) map[string]any {
	result := map[string]any{}
	for _, layer := range layers {
		if layer.Value == nil {
			continue
		}
		result = mergeInto(result, layer.Value, o, "")
	}
	return result
}

// applyStrategy implements SPEC section 3.2.
//
// Under first-match only the highest-precedence root that contributed a file survives - the
// lower roots are dropped from the merge *and* from sources, because the option means "the
// others were never consulted", not "the others lost". The rootless layers - defaults, env,
// overrides - always survive: the option scopes the file layers only.
func applyStrategy(layers []Layer, strategy string) []Layer {
	if strategy != "first-match" {
		return layers
	}
	winning := ""
	for _, layer := range layers {
		if layer.Root != "" {
			winning = layer.Root
		}
	}
	kept := make([]Layer, 0, len(layers))
	for _, layer := range layers {
		if layer.Root == "" || layer.Root == winning {
			kept = append(kept, layer)
		}
	}
	return kept
}
